src/fetch/payouts.py

```diff
@@ -0,0 +1,348 @@
+
+    def total_outgoing_eth(self) -> int:
+        """Total outgoing amount (including slippage) for the payout."""
+        return self.payment_eth + self.secondary_reward_eth + self.slippage_eth
+
+    def total_cow_reward(self) -> int:
+        """Total outgoing COW token reward"""
+        return self.reward_cow + self.secondary_reward_cow
+
+    def total_eth_reward(self) -> int:
+        """Total outgoing COW token reward"""
+        return self.payment_eth - self.exec_cost + self.secondary_reward_eth
+
+    def is_overdraft(self) -> bool:
+        """
+        True if the solver's complete combined data results in a net negative
+        """
+        return self.total_outgoing_eth() < 0
+
+    def as_payouts(self) -> list[Transfer]:
+        """
+        Isolating the logic of how solvers are paid out according to their
+            execution costs, rewards and slippage
+        """
+        # We do not handle overdraft scenario here!
+        assert not self.is_overdraft()
+        reward_eth = int(self.total_eth_reward())
+        reward_cow = int(self.total_cow_reward())
+
+        reimbursement_eth = int(self.exec_cost + self.slippage_eth)
+        # We do not have access to token conversion here, but we do have other converted values
+        # x_eth:x_cow = y_eth:y_cow --> y_cow = y_eth * x_cow / x_eth
+        reimbursement_cow = (
+            (reimbursement_eth * reward_cow) // reward_eth if reward_eth != 0 else 0
+        )
+
+        if reimbursement_eth > 0 > reward_cow:
+            # If the total payment is positive but the total rewards are negative,
+            # pay the total payment in ETH. The total payment corresponds to reimbursement,
+            # reduced by the negative reward.
+            # That assertion was not necessary because:
+            # reimbursement_eth + reward_eth
+            # = self.total_eth_reward() + self.exec_cost + self.slippage_eth
+            # = self.payment_eth + self.secondary_reward_eth + self.slippage_eth
+            # = self.total_outgoing_eth()
+            # >= 0 (because not self.is_overdraft())
+            return (
+                [
+                    Transfer(
+                        token=None,
+                        recipient=self.solver,
+                        amount_wei=reimbursement_eth + reward_eth,
+                    )
+                ]
+                if reimbursement_eth + reward_eth > 0
+                else []
+            )
+
+        if reimbursement_eth < 0 < reward_cow:
+            # If the total payment is positive but the total reimbursement is negative,
+            # pay the total payment in COW. The total payment corresponds to a payment of rewards,
+            # reduced by the negative reimbursement.
+            return [
+                Transfer(
+                    token=Token(COW_TOKEN_ADDRESS),
+                    recipient=self.solver,
+                    amount_wei=reimbursement_cow + reward_cow,
+                )
+            ]
```

Comment:
do we need the
```python
                if reimbursement_eth + reward_eth > 0
                else []
```
here as well? just to rule out the case that `reimbursement_eth + reward_eth = 0`.

Reply:
Perhaps. Although we have a try catch here don't we?

Reply:
Fixed.